Interface functionality based on MicroUI - Initial wrapper

WhyPenguins opened the pull request:
# Description

This PR adds (initial and WIP) interface functionality to SplashKit, by integrating the MicroUI library into the input/rendering code and providing some basic functionality via wrappers.

 - includes MicroUI in the external repository + build process
 - hooks MicroUI into input events (part of `process_events`) and rendering (separate `draw_interface` function)
 - wraps the following functionality:
   - windows (referred to as 'panels')
   - popups
   - labels
   - buttons
 - setting interface font/font size
 - adds UI test 


## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as
      expected)
- [ ] Documentation (update or new)

# How Has This Been Tested?

- Compiled and tested under Windows
- UI test has been run, and functionality such as rendering of windows, and responding to input such as mouse clicks/drags have been observed.
- Keyboard input has also been tested via calls to (currently unwrapped) text input elements. Requires `start_reading_text` to be called beforehand.

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
  - I intend to once the API has been mostly settled on
- [x] My changes generate no new warnings
- [ ] I have requested a peer review

WhyPenguins commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Add simple initial wrapper of MicroUI](https://trello.com/c/lNjr7ZGf/457-add-simple-initial-wrapper-of-microui)

hugglesfox commented on coresdk/src/backend/interface_driver.cpp:
```diff
@@ -0,0 +1,256 @@
+    int _text_width(mu_Font font, const char *text, int len)
+    {
+        if (len == -1) { len = strlen(text); }
+        return text_width((std::string)text, FONT_NAME, FONT_SIZE);
```
Could the font name and size come from mu_Font? (unfortunately in the mu source it's defined as a void pointer so whether that's possible depends on what it is at runtime)

WhyPenguins replied:
Possibly, in fact I was hoping to make this work like that. In later commits I just have two global variables that the user can set with functions, but it would indeed be better as you say. Will give it a go and backport it to this PR too.

WhyPenguins replied:
The initial improvement commit for this is 5a2286f9ab93bde233ed47b51013ee31232ee64c, and after that I've made another commit that uses the mu_Font properly (d8e8a33c62a6c028330da82e5a52c409136f6526). Works well, can even change font for just parts of the interface now 😁

hugglesfox commented on coresdk/src/backend/interface_driver.h:
```diff
@@ -0,0 +1,42 @@
+
+// temporary hash defines
+#define FONT_SIZE 12
+#define FONT_NAME font_named("input")
```
Should these be here?

hugglesfox replied:
Op sorry just saw the note in the PR description.

hugglesfox commented on coresdk/src/backend/interface_driver_atlas.h:
```diff
@@ -0,0 +1,997 @@
+
+rectangle atlas[ATLAS_FONT+128];
+
+void _initialize_atlas_map(){
```
Could atlas be initialised at compile time?

hugglesfox commented on coresdk/src/backend/interface_driver_atlas.h:
```diff
@@ -0,0 +1,997 @@
+};
+
+
+rectangle atlas[ATLAS_FONT+128];
```
Should this be static?

hugglesfox commented on coresdk/src/backend/interface_driver_atlas.h:
```diff
@@ -0,0 +1,997 @@
+enum { ATLAS_WIDTH = 128, ATLAS_HEIGHT = 128 };
+
+
+static unsigned char atlas_texture[ATLAS_WIDTH * ATLAS_HEIGHT] = {
```
I'm guessing this is a bin dump of an icon pack? Might be nice to add a comment and cite the source

WhyPenguins replied:
Yep, this comes straight from the MicroUI demo. The original file doesn't have attribution info and I forgot to copy it from one of the other files, will fix!

WhyPenguins replied:
Added attribution here - 06f9c6c5d454c023a0f8de614fcb0f99c19dd806

hugglesfox commented on coresdk/src/coresdk/interface.cpp:
```diff
@@ -0,0 +1,62 @@
+    void label(const string& label)
+    {
+        sk_interface_label(label);
+    }
```
Add whitespace after the function

hugglesfox commented on coresdk/external:
This will also need to be added to the splashkit-external repo

hugglesfox reviewed (changes requested):
Well done. Just a few questions and suggestions.

WhyPenguins commented on coresdk/src/backend/interface_driver.cpp:
```diff
@@ -0,0 +1,256 @@
+        return text_height(alphabet, FONT_NAME, FONT_SIZE);
+    }
+
+    void _initialize_button_and_key_map()
```
Indeed, the original C code this is based on (the MicroUI demo) used designated initializers to handle this at compiler time - maybe something similar can be done in C++? Will look into it!

WhyPenguins replied:
Looked into it but seems like it isn't that easy. Considering the initialization takes less than a hundredth of a millisecond I didn't want to spend too long on it. Hopefully that's okay 😅

WhyPenguins commented on coresdk/src/backend/interface_driver.cpp:
```diff
@@ -0,0 +1,256 @@
+
+    void sk_interface_init()
+    {
+        if (ctx == nullptr)
```
Yep sounds good. Just remnants of a previous design I should have removed 😅

WhyPenguins replied:
Looking at the other initialization functions in SplashKit, it seems they generally just don't check 😅  I can't find any asserts at all in the SplashKit code base (aside from one of the tests), so just in case there's a reason for that I've opted to just remove the check (in commit 3557972dc3dac287ba73dc57377c70234a679e7a). Seem alright?

WhyPenguins commented on coresdk/src/backend/interface_driver.cpp:
```diff
@@ -0,0 +1,256 @@
+            ctx->text_height = _text_height;
+
+            ui_atlas = sk_create_bitmap(ATLAS_WIDTH, ATLAS_HEIGHT);
+            for(unsigned int i = 0; i < 1/*_sk_num_open_windows*/; i++)
```
From what I could tell it really _should_ be `_sk_num_open_windows`, but that's static to the graphics driver's unit and there was no way to access it. Have fixed in a later commit where I don't change the SDL_Texture directly.

WhyPenguins commented on coresdk/src/backend/interface_driver.cpp:
```diff
@@ -0,0 +1,256 @@
+            {
+                // sk_create_bitmap created SDL_Texture is SDL_PIXELFORMAT_RGBA8888, while atlas_texture only has 1 channel.
+                // Cannot use SDL_LockTexture/SDL_UnlockTexture because the SDL_Texture does not have SDL_TEXTUREACCESS_STREAMING enabled
+                // So create a temporary rgba copy of the data, update the texture, then tidy up.
```
Definitely an option, but in a later commit I just ended up using `sk_draw_pixel` for each pixel instead, which avoids the allocation and simplifies the logic a bit.

I doubt it's the most efficient way, but to do this properly would require adding another function to the graphics driver, as not all the needed info can be accessed outside of it. Would that be preferred?

hugglesfox replied:
I would have to see the pixel implementation to be sure but my guess is that it would be nicer if we could use a surface, that way the GPU can handle drawing the pixels rather then having to wait for the CPU to feed it a pixel at a time.

Adding a function to the graphics driver sounds like it might be the best option to me?

WhyPenguins replied:
Sounds good. The original `sk_draw_pixel` implementation is in commit 16fe189c631debe145d819152d921607b872edb6 (took ~55ms) and the improved one is here 76d59e977aef36c4008b71c61057438fc0efc254 (takes ~1ms).

I ended up adding two functions to graphics_driver:
 - `sk_set_bitmap_pixel`, which ensures the `sk_drawing_surface` has a surface we can write to, then writes to it.
 - `sk_refresh_bitmap`, which then replaces the `sk_drawing_surface`'s textures with ones created from that surface.

WhyPenguins commented on coresdk/src/backend/interface_driver.cpp:
```diff
@@ -0,0 +1,256 @@
+
+        surface = to_surface_ptr(opts.dest);
+
+        if (surface)
```
Wondered the same thing, but most of the code base seems to prefer to log and continue rather than exit (and `to_surface_ptr` informs the user that the surface doesn't exist, so they are aware at least). With that said I can make it assert, what would be preferable here?

hugglesfox replied:
Fair point. My preference is to always use an assert for when the error can only occur if the programmer stuffed up. However as the rest of library simply logs and continues then keeping consistent with that would be the better option.

WhyPenguins commented on coresdk/src/backend/interface_driver.cpp:
```diff
@@ -0,0 +1,256 @@
+
+    bool sk_interface_start_panel(const string& name, rectangle initial_rectangle)
+    {
+        bool open = mu_begin_window(ctx, name.c_str(), to_mu(initial_rectangle));
```
Ah this isn't an exception, it's the library returning whether the panel (or 'window') is open or not - they can be closed as well. Same thing for popups.

WhyPenguins commented:
@hugglesfox Thanks so much for reviewing this! Some of these issues are fixed up in later commits, will try and backport the changes to this PR and fix up the remaining issues as well.

hugglesfox commented on coresdk/src/backend/interface_driver.cpp:
```diff
@@ -0,0 +1,256 @@
+    int _text_height(mu_Font font)
+    {
+        // We don't recieve a string, so use a string that contains A-Z and a-z
+        const char* alphabet = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz";
```
> Regarding characters other than Latin1, the rest of SplashKit generally works alright actually, and I've been making patches for the parts that don't quite support it.

Ah sorry. I based that take off the [sk_text_size](https://github.com/thoth-tech/splashkit-core/blob/31b5afa890e4e4db8380d461fc602653ae66990f/coresdk/src/backend/text_driver.cpp#L180-L194) function which uses [TTF_SizeText](https://wiki.libsdl.org/SDL2_ttf/TTF_SizeText) which only supports Latin1. If other areas of the code base support non latin characters then I'll add swapping to [TTF_SizeUTF8](https://wiki.libsdl.org/SDL2_ttf/TTF_SizeUTF8) to the trello board.

Looking into it, the height of the font doesn't depend on the text input. I think it might be a limitation imposed by the splashkit api. We could update the splashkit api to use [TTF_FontHeight](https://wiki.libsdl.org/SDL2_ttf/TTF_FontHeight) for getting the font height which doesn't require the text to be known.

WhyPenguins replied:
Haha I'd already found that issue and added it to the Trello board at the time 🤣 , just need to get around to actually making the PR.
Regarding the height, in commit daffdb41164315c39676671b98196e40785fa48b I added a function to text_driver `sk_text_height` which just calls `TTF_FontHeight` internally, so now this callback uses that.

WhyPenguins commented on coresdk/src/backend/interface_driver.cpp:
```diff
@@ -0,0 +1,256 @@
+            {
+                switch (cmd->type)
+                {
+                    case MU_COMMAND_TEXT: sk_draw_text(surface, FONT_NAME, FONT_SIZE, cmd->text.pos.x, cmd->text.pos.y, cmd->text.str, from_mu(cmd->text.color));break;
```
Fixed this and other layout issues in c4b04d7b79ab4ac17d0fc7290f5f8e37e369477f -  hopefully alright now!

WhyPenguins commented:
Alright I think I've addressed all the issues! Hopefully now it's in a bit better shape - let me know if there are any other improvements that can be made 😁

hugglesfox reviewed (approved):
lgtm!
